### PR TITLE
Implement profile screen

### PR DESCRIPTION
**Pull request by lucyxubroad**
# Summary

Implement the profile screen for Flick.

# Detailed Changes

- Added ProfileViewController to contain Profile screen
- Added ListTableViewCell to contain the list previews in ProfileViewController
- Added UsersPreviewView to create users lined up in a row
 
# Addressed Issues

#8 

# Demonstration

profile

# Notes
- Still have to update the bottom sheet view for profile
- Some images seem pretty low quality to me
- I left notes here and there so take a look and let me know what you think?

### CODE REVIEW

File: Flick/Controllers/ProfileViewController.swift [hunk 1]
```diff
@@ -8,26 +8,283 @@
...
+        listsContainerView.addSubview(listsTableView)
+        view.addSubview(listsContainerView)
+
+        // TODO: Replace button image
+        addListButton.setImage(UIImage(named: "addButton"), for: .normal)
+        addListButton.layer.cornerRadius = addListButtonSize.width / 2
+        view.addSubview(addListButton)
+
+        setupConstraints()
+
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:160]: nit remove newline

File: Flick/Controllers/ProfileViewController.swift [hunk 2]
```diff
@@ -8,26 +8,283 @@
...
+    }
+
+    private func setupConstraints() {
+
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:164]: nit remove newline

File: Flick/Controllers/ProfileViewController.swift [hunk 3]
```diff
@@ -8,26 +8,283 @@
...
+        // Calculate width of friends preview based on number of friends and spacing between cells
+        let numFriendsInPreview = min(friends.count, 7) + 1
+        let fullFriendsWidth = numFriendsInPreview * 20
+        let overlapFriendsWidth = (numFriendsInPreview-1) * condensedCellSpacing * -1
+        let friendsPreviewWidth = fullFriendsWidth - overlapFriendsWidth
+
+        let padding = 20
+
+        usernameLabel.sizeToFit()
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:173]: could we move this to viewDidLoad?

File: Flick/Controllers/ProfileViewController.swift [hunk 4]
```diff
@@ -8,26 +8,283 @@
...
+            make.centerX.equalToSuperview()
+            make.height.equalTo(20)
+            make.width.equalTo(userInfoViewWidth)
+        }
+
+        activitySummaryCollectionView.snp.makeConstraints { make in
+            make.top.equalTo(userInfoView.snp.bottom).offset(18)
+            make.centerX.equalToSuperview()
+            make.width.equalTo(146) // Temp
+            make.height.equalTo(40)
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:212]: the g in Thoughts getting cut off might also have to do with this height

File: Flick/Controllers/ProfileViewController.swift [hunk 5]
```diff
@@ -8,26 +8,283 @@
...
         }
+
+        notificationButton.snp.makeConstraints { make in
+            make.size.equalTo(sideButtonsSize)
+            make.top.equalTo(settingsButton)
+            make.trailing.equalTo(settingsButton.snp.leading).offset(-7)
+        }
+
+        listsContainerView.snp.makeConstraints { make in
+            make.top.equalTo(activitySummaryCollectionView.snp.bottom).offset(22) //200 is temp
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:228]: you can prob remove the comment?

File: Flick/Controllers/ProfileViewController.swift [hunk 6]
```diff
@@ -8,26 +8,283 @@
...
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: activitySummaryCellReuseIdentifier, for: indexPath) as? SummaryCollectionViewCell else { return UICollectionViewCell() }
+        cell.configure(with: activitySummary[indexPath.item])
+        return cell
+    }
+}
+
+extension ProfileViewController: UICollectionViewDelegateFlowLayout {
+    
+
```
Comment (haiyingweng) [inline Flick/Controllers/ProfileViewController.swift:268]: remove one of the newlines

File: Flick/Views/UsersPreviewView.swift [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    /// Sets usersPreview to first numMaxusers users in array if number of users in array exceeds
+    /// numMaxusers, otherwise sets usersPreview to users
+    func getUsersPreview() {
+        usersPreview = Array(users.prefix(numMaxusers))
+    }
+
+    func setupViews() {
+
```
Comment (haiyingweng) [inline Flick/Views/UsersPreviewView.swift:54]: remove newline

File: Flick/Views/UsersPreviewView.swift [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        let usersLayout = UICollectionViewFlowLayout()
+        usersLayout.minimumInteritemSpacing = cellSpacing
+
+        usersCollectionView = UICollectionView(frame: .zero, collectionViewLayout: usersLayout)
+        usersCollectionView.delegate = self
+        usersCollectionView.dataSource = self
+        usersCollectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: usersCellReuseIdentifier)
+        usersCollectionView.backgroundColor = .none
+        addSubview(usersCollectionView)
+
```
Comment (haiyingweng) [inline Flick/Views/UsersPreviewView.swift:64]: remove newline

File: Flick/Views/UsersPreviewView.swift [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        usersCollectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: usersCellReuseIdentifier)
+        usersCollectionView.backgroundColor = .none
+        addSubview(usersCollectionView)
+        
+    }
+
+    func setupConstraints() {
+        usersCollectionView.snp.makeConstraints { make in
+            make.edges.equalToSuperview()
+        }
```
Comment (haiyingweng) [inline Flick/Views/UsersPreviewView.swift:70]: could prob just move this into `setupViews`

File: Flick/Views/UsersPreviewView.swift [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    // MARK: - Private View Vars
+    private var usersCollectionView: UICollectionView!
+
+    // MARK: - Private Data Vars
+    private var cellSpacing: CGFloat!
+    private var modeCellSpacing: [UsersLayoutMode : CGFloat] = [
+        .friends : -8,
+        .collaborators : -5
+    ]
+    private let numMaxusers = 7
```
Comment (haiyingweng) [inline Flick/Views/UsersPreviewView.swift:25]: `numMaxUsers`

File: Flick/Views/UsersPreviewView.swift [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+extension UsersPreviewView: UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        switch usersLayoutMode {
+        case .collaborators:
+            return usersPreview.count
+        case .friends:
+            return usersPreview.count + 1 // Add one more cell for last detail cell
+        case .none:
+            return usersPreview.count
```
Comment (haiyingweng) [inline Flick/Views/UsersPreviewView.swift:84]: confused why we don't do +1 for collaborators

File: Flick/Cells/ListTableViewCell.swift [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    func setupCollaborators(collaborators: [String]) {
+
+        let collaboratorsPreviewView = UsersPreviewView(users: collaborators, usersLayoutMode: .collaborators)
+        addSubview(collaboratorsPreviewView)
+
+        // Calculate width of friends preview based on number of friends and spacing between cells
+        let numCollaborators = collaborators.count
+        let fullCollaboratorsWidth = numCollaborators * 20
+        let overlapCollaboratorsWidth = (numCollaborators-1) * collaboratorsCellSpacing * -1
+        let collaboratorsPreviewWidth = fullCollaboratorsWidth - overlapCollaboratorsWidth
```
[the remark this answers is not in the document]
Comment (lucyxubroad) [reply Flick/Cells/ListTableViewCell.swift:66]: Oh hmm.. very interesting. I didn't think of that. I'll ask Cindy.

[the remark this answers is not in the document]
Comment (haiyingweng) [reply Flick/Cells/ListTableViewCell.swift:66]: Yes bc when we show a media list, there’s a view that shows you the collaborators preview

**Pull request merged**